### ISSUE TITLE
hello from Intake

### ISSUE THREAD

**Issue by martindurant** [Gijs-Koot/cbs-intake#1]
I must admit, I don't even know what CDS data is, but if you would like some help in getting this project off the ground, please be in touch!

**Comment by Gijs-Koot** [comment]
Hello Martin, that's very friendly. I could indeed use some help. CBS is a Dutch govermental institution that gathers data and statistics, the name is short for Central Agency for Statistics. 

They publish a large number of datasets (~7000), on topics such as demographics, locations, education, etcetera. They maintain an API following the odata v3 protocol. That's a pretty old, not well-supported protocol, in particular, there is no python package that supports it, I think. It's quite simple to work with. 

There is a feed that lists all the tables included at 'https://opendata.cbs.nl/ODataCatalog/Tables'. This links a 19MB xml file with descriptions and links to the underlying data. 

- A script that creates a up-to-date catalog (a .yml file) based on their api
  - If this package is to be published as a datapackage, it should probably include the most up-to-date version of the official catalog as part of a build process (any thoughts on this?)
- A `CBSOdataDriver` that reads datasets. The published format is very untidy, with unyieldy hierarchical columns and rows, and multiple tables that are joined together in unpractical ways to work with them. I'm considering scripting the logic to untangle these datasets into 'tidy' dataframes in the driver. That would be a rather opiniated piece of code, and I'm not sure if it's a good idea to include that in the driver, or to perhaps create multiple drivers for 'raw' access and 'tidy' access. 

I have a plan and am looking to work on this in the coming weeks or months. Perhaps from the description above you can give me a quick opinion on whether this looks to be a usecase for the 'intake' package, and how you think the particularities of this situation (a dynamic, rather large catalog, an involved transformation of the data) should be handled in the framework.

**Comment by martindurant** [comment]
Yes certainly, I think Intake can do something like this. Perhaps https://github.com/CityOfLosAngeles/intake-dcat is a nice motivating case for public data perhaps of a different sort (cc @ian-r-rose , in case you know anything about the data mentioned here).

Intake has seen both patterns of catalog generation: 
- creating static YAML prescriptions out of some source, so that you need to update it time-to-time; updating could be in-place, or versioned in a git repo and/or conda package; 
- dynamic catalogs which query some service on-demand and have data source entries only ephemerally, in memory

It sounds like you are somewhere in-between, with a large set of things, which takes some time and complexity to scan. Maybe you would want the list of entries to update only occasionally. I would recommend either writing your own logic for that, or having users call the `.persist()`method of the catalog, which allows for a local copy to be made with a given expiry time.

**Comment by martindurant** [comment]
> That would be a rather opiniated piece of code, and I'm not sure if it's a good idea to include that in the driver, or to perhaps create multiple drivers for 'raw' access and 'tidy' access.

It is hard for me to say from the outside. Since the code is messy, it's probably fine to have opinions on what users are likely to want. You could have separate drivers or options within the same driver, just matter of choice. Note that you can list multiple drivers to the same data-source (so long as the output type is the same), and you can also provide "UserParameters" in catalog specs to indicate the options users should consider.

**Comment by ian-r-rose** [comment]
I don't know much about CBS data either, but I'd be happy to make myself available to bounce ideas off of while you are working on this @Gijs-Koot, since it seems like there are some real similarities in intent between this and `intake-dcat`.

I also made some opinionated decisions in that package, particularly around which drivers to open a given dataset with if there are multiple options. So far that seems to be going okay, and most users don't really seem to care about that as long as they can access the data.

**Comment by martindurant** [comment]
> most users don't really seem to care about that as long as they can access the data

The Intake mantra! :)

**Comment by Gijs-Koot** [comment]
Thank you both for your input and the `intake-dcat` link, I will have a closer look at that. I will reach out for help if I've made some progress!

**Comment by martindurant** [comment]
Did you manage to make any progress, anything I can help with?

**Comment by Gijs-Koot** [comment]
Hi Martin, thanks for your interest. I'm sorry to say I haven't made any
progress, but I just had a new addition to the family, which is great, and
this is not really a priority at the moment. I do appreciate your efforts,
at the moment though, they're better spent on other project related to the
Intake package. Heard you talk on the __init__ podcast recently, really
nice and I'll be sure to recommend some fellow data scientists to look into
intake.
